sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -630,4 +631,35 @@ object PartitioningUtils extends SQLConfHelper{
     case (DoubleType, LongType) | (LongType, DoubleType) => StringType
     case (t1, t2) => TypeCoercion.findWiderTypeForTwo(t1, t2).getOrElse(StringType)
   }
+
+  /**
+   * Evaluating partition filter for datasource scan at runtime.
+   * This is mainly used to evaluate partition conditions during task runtime and
+   * complete further pruning of push-down dataFilter, which can reduce datasource scan IO.
+   * @param partitionSchema schema of partition.
+   * @param partitionValues partition values.
+   * @param partitionFilter partition filter.
+   * @return the result of evaluate result.
+   */
+  def evaluatePartitionFilter(
+      partitionSchema: StructType,
+      partitionValues: Option[InternalRow],
+      partitionFilter: sources.Filter): Boolean = {
+    val predicate =
+      StructFilters.filterToExpression(partitionFilter, StructFilters.toRef(partitionSchema))
+    if (predicate.isDefined) {
```

Comment:
Maybe we can extract line651 ~ 655 into a new function, then return
```scala
...
predicate.isEmpty || predicate.isDefined && partitionValues.exists(newFunc.eval)
```

Is that right?